### PR TITLE
AzureResourceGroupDeployment: Improve docs/error logs to surface SPN details to user

### PR DESCRIPTION
**Pull request by issacnitin**

### CODE REVIEW

File: Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts [hunk 1]
```diff
@@ -18,7 +19,27 @@ export class DeploymentScopeBase {
     }
 
     public async deploy(): Promise<void> {
-        await this.createTemplateDeployment();
+        try {
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts]: the try-catch should be moved inside createTemplateDeployment(). This way you won;t have to handle it in two places.

File: Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts [hunk 2]
```diff
@@ -18,7 +19,27 @@ export class DeploymentScopeBase {
...
+            await this.createTemplateDeployment();
+        } catch (error) {
+            if((error as string).toLowerCase().indexOf("serviceprincipal") != -1) {
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts]: is it not possible to have a better string to detect - 'serviceprincipal' is too generic. Will there not be an error code or status or something which indicates 403 or access denied or forbidden etc.
Can you please point me to a sample failure.

File: Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts [hunk 3]
```diff
@@ -18,7 +19,27 @@ export class DeploymentScopeBase {
...
+                try {
+                    var spnName = await this.getServicePrincipalName();
+                    tl.warning(tl.loc("ServicePrincipalRoleAssignmentDetails", spnName, this.taskParameters.resourceGroupName));
+                } 
+                catch (err)
+                {
+                    tl.error(err);
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts]: Add some prefix to error to indicate that the error occurred while getting SPN name. And log this message only for debug mode `tl.debug()`

**Comment by bishal-pdMSFT** [comment]
Please make same change for v2 task as well.

File: Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts (hunk 1 above)
Comment (issacnitin) [reply Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts]: Done

**Comment by issacnitin** [comment]
> Please make same change for v2 task as well.

Done

File: Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts (hunk 2 above)
Comment (issacnitin) [reply Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts]: I've updated to check for status code. 

Sample Failure:
2019-12-04T10:00:50.9956997Z ##[error]The client '94405baa-4126-47a1-baeb-e5ae9c4151e1' with object id '94405baa-4126-47a1-baeb-e5ae9c4151e1' does not have authorization to perform action 'Microsoft.Resources/deployments/write' over scope '/subscriptions/c00d16c7-6c1f-4c03-9be1-6934a4c49682/resourcegroups/issacnitinrg/providers/Microsoft.Resources/deployments/azuredeploy-20191204-153050-957f' or the scope is invalid. If access was recently granted, please refresh your credentials.

File: Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts (hunk 3 above)
Comment (issacnitin) [reply Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts]: Done

File: Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts [hunk 4]
```diff
@@ -534,8 +534,13 @@ export class ResourceGroup {
                 armClient.deployments.createOrUpdate(this.taskParameters.deploymentName, deployment, (error, result, request, response) => {
                     if (error) {
                         if(error.code == "ResourceGroupNotFound" && retryCount > 0){
-                            return this.waitAndPerformAzureDeployment(armClient, deployment, retryCount);
+                            return this.waitAndPerformAzureDeployment(armClient, deployment, retryCount, spnName);
+                        }
+                        
+                        if(error.statusCode == 403) {
+                            tl.error(tl.loc("ServicePrincipalRoleAssignmentDetails", spnName, this.taskParameters.resourceGroupName));
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts]: why pass it as an argument through more than one functions? better to just call this.getServicePrincipalName() here.

**Comment by bishal-pdMSFT** [comment]
please also port it to M161 branch

File: Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts (hunk 4 above)
Comment (issacnitin) [reply Tasks/AzureResourceGroupDeploymentV2/operations/ResourceGroup.ts]: To do that we'll have to make the callback Async, which won't work

**Pull request merged**